File: iosApp/iosApp/Modules/CurrencySwap/CurrencySwapInteractor.swift
// Created by web3d4v on 14/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

struct SwapData {
    let inputCurrency: Currency
    let outputCurrency: Currency
    let inputAmount: BigInt
}

protocol CurrencyInteractorLister: AnyObject {
    func handle(swapEvent event: UniswapEvent)
}

enum CurrencySwapInteractorOutputAmountState {
    case loading
    case ready
}

enum CurrencySwapInteractorApprovalState {
    case approve
    case approving
    case approved
}

enum CurrencySwapInteractorSwapState {
    case noPools
    case notAvailable
    case swap
}

protocol CurrencySwapInteractor: AnyObject {
    var outputAmount: BigInt { get }
    var outputAmountState: CurrencySwapInteractorOutputAmountState { get }
    var approvingState: CurrencySwapInteractorApprovalState { get }
    var swapState: CurrencySwapInteractorSwapState { get }
    var swapService: UniswapService { get }
    func defaultCurrencyFrom(for network: Network) -> Currency
    func defaultCurrencyTo(for network: Network) -> Currency
    func networkFees(network: Network) -> [NetworkFee]
    func getQuote(data: SwapData)
    func isCurrentQuote(data: SwapData) -> Bool
    func addListener(_ listener: CurrencyInteractorLister)
    func removeListener(_ listener: CurrencyInteractorLister)
    func approveSwap(
        network: Network,
        currency: Currency,
        password: String,
        salt: String
    )
    func balance(currency: Currency, network: Network) -> BigInt
    func fiatPrice(currency: Currency) -> Double
}

final class DefaultCurrencySwapInteractor {
    private let network: Network
    private let walletService: WalletService
    private let networksService: NetworksService
    let swapService: UniswapService
    private let currencyStoreService: CurrencyStoreService
    
    private var listener: WeakContainer?
    
    init(
        network: Network,
        walletService: WalletService,
        networksService: NetworksService,
        swapService: UniswapService,
        currencyStoreService: CurrencyStoreService
    ) {
        self.network = network
        self.walletService = walletService
        self.networksService = networksService
        self.swapService = swapService
        self.currencyStoreService = currencyStoreService
        configureUniswapService()
    }
    
    deinit {
        print("[DEBUG][Interactor] deinit \(String(describing: self))")
    }
}

extension DefaultCurrencySwapInteractor: CurrencySwapInteractor {

    var outputAmount: BigInt { swapService.outputAmount }
    
    var outputAmountState: CurrencySwapInteractorOutputAmountState {
        switch swapService.outputState {
        case is OutputState.Loading: return .loading
        default: break
        }
        switch swapService.poolsState {
        case is PoolsState.Loading: return .loading
        default: break
        }
        return .ready
    }
    
    var approvingState: CurrencySwapInteractorApprovalState {
        switch swapService.approvalState {
        case is ApprovalState.NeedsApproval: return .approve
        case is ApprovalState.Approving: return .approving
        default: return .approved
        }
    }
    
    var swapState: CurrencySwapInteractorSwapState {
        guard !(swapService.poolsState is PoolsState.NoPoolsFound) else { return .noPools }
        switch swapService.outputState {
        case is OutputState.Unavailable: return .notAvailable
        default: return .swap
        }
    }
    
    func defaultCurrencyFrom(for network: Network) -> Currency {
        walletService.currencies(network: network)[safe: 0] ?? network.nativeCurrency
    }
    func defaultCurrencyTo(for network: Network) -> Currency {
        walletService.currencies(network: network)[safe: 1] ?? network.nativeCurrency
    }
    
    func networkFees(network: Network) -> [NetworkFee] {
        networksService.networkFees(network: network)
    }

    func getQuote(data: SwapData) {
        swapService.inputAmount = data.inputAmount
        swapService.inputCurrency = data.inputCurrency
        swapService.outputCurrency = data.outputCurrency
        let input = data.inputAmount.toDecimalString()
        let inputC = data.inputCurrency.symbol
        let outputC = data.outputCurrency.symbol
        print("[SWAP][QUOTE][REQUEST] - Swap: \(input) '\(inputC)' to '\(outputC)'")
    }
    
    func isCurrentQuote(data: SwapData) -> Bool {
        guard swapService.inputAmount == data.inputAmount else { return false }
        guard swapService.inputCurrency.symbol == data.inputCurrency.symbol else { return false }
        guard swapService.outputCurrency.symbol == data.outputCurrency.symbol else { return false }
        return true
    }

    
    func approveSwap(
        network: Network,
        currency: Currency,
        password: String,
        salt: String
    ) {
        // TODO: Review scenarios where this can fail or return without triggering requestApproval
        guard let wallet = networksService.wallet(network: network) else { return }
        do {
            try wallet.unlock(password: password, salt: salt)
            swapService.requestApproval(
                currency: currency,
                wallet: wallet,
                completionHandler: { _ in }
            )
        } catch {
            // do nothing
        }
    }
    
    func balance(currency: Currency, network: Network) -> BigInt {
        walletService.balance(network: network, currency: currency)
    }
    
    func fiatPrice(currency: Currency) -> Double {
        currencyStoreService.marketData(currency: currency)?.currentPrice?.doubleValue ?? 0
    }
}

private extension DefaultCurrencySwapInteractor {
    
    func configureUniswapService() {
        guard let wallet = networksService.wallet(network: network) else {
            fatalError("Unable to configure Uniswap service, no wallet found.")
        }
        let provider = networksService.provider(network: network)
        swapService.wallet = wallet
        swapService.provider = provider
    }
}

extension DefaultCurrencySwapInteractor: UniswapListener {
    
    func addListener(_ listener: CurrencyInteractorLister) {
        self.listener = WeakContainer(listener)
        swapService.add(listener___: self)
    }
    
    func removeListener(_ listener: CurrencyInteractorLister) {
        self.listener = nil
        swapService.remove(listener___: self)
    }
    
    func handle(event___ event: UniswapEvent) {
        print("[SWAP][EVENT] - \(event)")
        emit(event)
    }

    private func emit(_ event: UniswapEvent) {
        listener?.value?.handle(swapEvent: event)
    }

    private class WeakContainer {
        weak var value: CurrencyInteractorLister?

        init(_ value: CurrencyInteractorLister) {
            self.value = value
        }
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/currencySwap/CurrencySwapInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.currencySwap

import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.services.uniswap.ApprovalState
import com.sonsofcrypto.web3lib.services.uniswap.OutputState
import com.sonsofcrypto.web3lib.services.uniswap.PoolsState
import com.sonsofcrypto.web3lib.services.uniswap.UniswapEvent
import com.sonsofcrypto.web3lib.services.uniswap.UniswapListener
import com.sonsofcrypto.web3lib.services.uniswap.UniswapService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.types.Currency
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.legacy.NetworkFee
import com.sonsofcrypto.web3lib.types.BigInt
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapInteractorApprovalState.APPROVE
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapInteractorApprovalState.APPROVED
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapInteractorApprovalState.APPROVING
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapInteractorOutputAmountState.LOADING
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapInteractorOutputAmountState.READY
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapInteractorSwapState.NOT_AVAILABLE
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapInteractorSwapState.NO_POOLS
import com.sonsofcrypto.web3walletcore.modules.currencySwap.CurrencySwapInteractorSwapState.SWAP

interface CurrencyInteractorLister {
    fun handle(event: UniswapEvent)
}

enum class CurrencySwapInteractorOutputAmountState { LOADING, READY }

enum class CurrencySwapInteractorApprovalState { APPROVE, APPROVING, APPROVED }

enum class CurrencySwapInteractorSwapState { NO_POOLS, NOT_AVAILABLE, SWAP }

interface CurrencySwapInteractor {
    val outputAmount: BigInt
    val outputAmountState: CurrencySwapInteractorOutputAmountState
    val approvingState: CurrencySwapInteractorApprovalState
    val swapState: CurrencySwapInteractorSwapState
    fun swapService(): UniswapService
    fun defaultCurrencyFrom(network: Network): Currency
    fun defaultCurrencyTo(network: Network): Currency
    fun networkFees(network: Network): List<NetworkFee>
    fun getQuote(from: Currency, to: Currency, amount: BigInt)
    fun isCurrentQuote(from: Currency, to: Currency, amount: BigInt): Boolean
    @Throws(Throwable::class)
    suspend fun approveSwap(network: Network, currency: Currency, password: String, salt: String)
    fun balance(currency: Currency, network: Network): BigInt
    fun fiatPrice(currency: Currency): Double
    fun add(listener: CurrencyInteractorLister)
    fun remove(listener: CurrencyInteractorLister)
}

class DefaultCurrencySwapInteractor(
    private val walletService: WalletService,
    private val networksService: NetworksService,
    private val swapService: UniswapService,
    private val currencyStoreService: CurrencyStoreService
): CurrencySwapInteractor, UniswapListener {
    private var listener: CurrencyInteractorLister? = null

    override val outputAmount: BigInt get() = swapService.outputAmount

    override val outputAmountState: CurrencySwapInteractorOutputAmountState get() {
        return if (swapService.outputState is OutputState.Loading) LOADING
        else if (swapService.poolsState is PoolsState.Loading) LOADING
        else READY
    }

    override val approvingState: CurrencySwapInteractorApprovalState get() =
        when (swapService.approvalState) {
            is ApprovalState.NeedsApproval -> APPROVE
            is ApprovalState.Approving -> APPROVING
            else -> APPROVED
        }

    override val swapState: CurrencySwapInteractorSwapState get() =
        if (swapService.poolsState is PoolsState.NoPoolsFound) { NO_POOLS }
        else if (swapService.outputState == OutputState.Unavailable) { NOT_AVAILABLE }
        else SWAP

    override fun swapService(): UniswapService = swapService

    override fun defaultCurrencyFrom(network: Network): Currency =
        walletService.currencies(network).getOrNull(0) ?: network.nativeCurrency

    override fun defaultCurrencyTo(network: Network): Currency =
        walletService.currencies(network).getOrNull(1) ?: network.nativeCurrency

    override fun networkFees(network: Network): List<NetworkFee> =
        networksService.networkFees(network)

    override fun getQuote(from: Currency, to: Currency, amount: BigInt) {
        swapService.inputCurrency = from
        swapService.outputCurrency = to
        swapService.inputAmount = amount
    }

    override fun isCurrentQuote(from: Currency, to: Currency, amount: BigInt): Boolean {
        if (swapService.inputAmount != amount) { return false }
        if (swapService.inputCurrency.symbol != from.symbol) { return false }
        if (swapService.outputCurrency.symbol != to.symbol) { return false }
        return true
    }

    @Throws(Throwable::class)
    override suspend fun approveSwap(
        network: Network, currency: Currency, password: String, salt: String
    ) {
        val wallet = networksService.wallet(network) ?: return
        wallet.unlock(password, salt)
        swapService.requestApproval(currency, wallet)
    }

    override fun balance(currency: Currency, network: Network): BigInt =
        walletService.balance(network, currency)

    override fun fiatPrice(currency: Currency): Double =
        currencyStoreService.marketData(currency)?.currentPrice ?: 0.toDouble()

    override fun add(listener: CurrencyInteractorLister) {
        this.listener = listener
        swapService.add(this)
    }

    override fun remove(listener: CurrencyInteractorLister) {
        this.listener = null
        swapService.remove(this)
    }

    override fun handle(event: UniswapEvent) {
        listener?.handle(event)
    }
}